docs/html/development/architecture/configuration-files.rst

```diff
@@ -0,0 +1,145 @@
+===========================
+Configuration File Handling
+===========================
+
+The ``pip._internal.configuration`` module is responsible for handling
+configuration files (eg. loading from and saving values to) that are used by
+pip. The module's functionality is largely exposed through and coordinated by
+the module's ``Configuration`` class.
+
+.. note::
+
+    This section of the documentation is currently being written. pip
+    developers welcome your help to complete this documentation. If you're
+    interested in helping out, please let us know in the
+    `tracking issue <https://github.com/pypa/pip/issues/6831>`_.
+
+
+.. _configuration-overview:
+
+Overview
+========
+
+pip stores configuration files in standard OS-appropriate locations, which are
+determined by ``appdirs``. These files are in the INI format and are processed
+with ``RawConfigParser``.
+
+pip uses configuration files in two operations:
+
+- During processing of command line options.
+  - Reading from *all* configuration sources
+- As part of ``pip config`` command.
+  - Reading from *all* configuration sources
+  - Manipulating a single configuration file
+
+Both of these operations utilize functionality provided the ``Configuration``
+object, which encapsulates all the logic for handling configuration files and
+provides APIs for the same.
+
+The remainder of this section documents the ``Configuration`` class, and
+discusses potential future refactoring ideas.
+
+
+.. _configuration-class:
+
+``Configuration`` class
+=======================
+
+``Configuration`` loads configuration values from sources in the local
+environment: a combination of configuration files and environment variables.
+
+It can be used in two "modes", for reading all the values from the local
+environment and for manipulating a single configuration file. It differentiates
+between these two modes using the ``load_only`` attribute.
```

Comment:
s/attribute/argument/?

The name `load_only` isn't entirely obvious to me, so maybe add some clarification here? Something like:

> The `load_only` argument specifies a single configuration file to load. If specified, the `Configuration` class can be used to read and edit values from that file. If `load_only` is `None`, all configuration sources in the local environment are read, and the interface is read-only.

Update: Reading below, `load_only` isn't a file, it's a *kind* of file. I'd make that point here.

Reply:
`attribute` is probably fine; I'll duplicate that sentence from later since it's pretty clear in communicating that.